File: config-auth-core-full-kt/assets/auth-core-full-template-kt/src/main/kotlin/com/example/auth/permission/entity/Permission.kt
package com.example.auth.permission.entity

import com.example.shared.domain.vo.DotSeparatedName
import com.example.shared.domain.vo.Id

data class Permission private constructor(
    val id: Id,
    val name: DotSeparatedName,
    val description: String? = null
) {
    companion object {
        fun create(id: String? = null, name: String, description: String? = null): Permission =
            tryCreate(id, name, description).getOrThrow()

        fun tryCreate(id: String? = null, name: String, description: String? = null): Result<Permission> {
            val validId = Id.tryCreate(id).getOrElse { return Result.failure(it) }
            val validName = DotSeparatedName.tryCreate(name).getOrElse { return Result.failure(it) }
            return Result.success(Permission(validId, validName, description))
        }
    }

    override fun equals(other: Any?): Boolean = other is Permission && other.id == this.id
    override fun hashCode(): Int = id.hashCode()
}

File: config-auth-core-full-cs/assets/auth-core-full-template-cs/Domain/Entities/Permission.cs
using Project.Shared.Kernel.Domain.Base;
using Project.Shared.Kernel.Domain.Results;

namespace Project.Auth.Domain.Entities;

public class Permission : Entity
{
    public string Key { get; private set; } = string.Empty;
    public string Description { get; private set; } = string.Empty;

    private Permission() { }

    public static Result<Permission> Create(string key, string description = "")
    {
        if (string.IsNullOrWhiteSpace(key))
            return Result<Permission>.Failure("Permission key is required.");

        if (!key.Contains(':'))
            return Result<Permission>.Failure("Permission key must follow 'resource:action' format (e.g. 'products:create').");

        return Result<Permission>.Success(new Permission
        {
            Id = Guid.NewGuid(),
            Key = key.Trim().ToLowerInvariant(),
            Description = description.Trim()
        });
    }
}
